### PR TITLE
Add extended job attributes and support job body marshalling

### PR DESCRIPTION
**Pull request by Napoleon0621**
#195

### CODE REVIEW

File: engine/redis_v2/model/job.proto [hunk 1]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package model;
+option go_package="github.com/lmstfy/engine/redis_v2/model";
+
+message ExtendedJob {
```
Comment (git-hulk) [inline engine/redis_v2/model/job.proto:5]: We can move this proto file into a dedicated dir and how about renaming it to `JobData` and body => data?

**Review by git-hulk — COMMENTED** [review]
Also need to encode body when shrinking into the secondary storage.

File: engine/redis_v2/model/job.proto (hunk 1 above)
Comment (Napoleon0621) [reply engine/redis_v2/model/job.proto:5]: Could you please specify "dedicated dir" ? 
 Do you mean under the project root directory?

File: engine/job.go [hunk 2]
```diff
@@ -31,6 +34,7 @@ type jobImpl struct {
 	ttl       uint32
 	delay     uint32
 	tries     uint16
+	pool      string
```
Comment (git-hulk) [inline engine/job.go:37]: Why do we need this field? I think it will be better to support passing the pool name in `AddJob`

Comment (Napoleon0621) [reply engine/job.go:40]: Ok, will add another input argument in AddJob function.

File: engine/job.go [hunk 3]
```diff
@@ -5,6 +5,9 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"errors"
+	"github.com/bitleak/lmstfy/engine/model"
```
Comment (git-hulk) [inline engine/job.go]: Please use goimports to fmt the import order

File: engine/job.go [hunk 4]
```diff
@@ -70,6 +77,21 @@ func NewJobWithID(namespace, queue string, body []byte, ttl uint32, tries uint16
 	}
 }
 
+func NewJobWithoutMarshal(namespace, queue string, body []byte, ttl, delay uint32, tries uint16, jobID string) Job {
```
Comment (git-hulk) [inline engine/job.go:77]: NewRawJob?

File: engine/job.go [hunk 5]
```diff
@@ -188,3 +211,14 @@ func (j *jobImpl) MarshalText() (text []byte, err error) {
...
+
+func marshalJobBody(body []byte) ([]byte, error) {
+	job := &model.JobData{
+		Data: body,
+	}
+	data, err := proto.Marshal(job)
+	if err != nil {
+		return nil, err
+	}
+	return data, err
```
Comment (git-hulk) [inline engine/job.go:223]:
```suggestion
	return proto.Marshal(job)
```

File: engine/redis/pool.go [hunk 6]
```diff
@@ -68,7 +71,12 @@ func (p *Pool) Get(namespace, queue, jobID string) (body []byte, ttlSecond uint3
 			ttl = 0
 		}
 		metrics.poolGetJobs.WithLabelValues(p.redis.Name).Inc()
-		return []byte(val), uint32(ttl), nil
+		res := &model.JobData{}
+		err = proto.Unmarshal([]byte(val), res)
+		if err != nil {
+			return nil, 0, err
```
Comment (git-hulk) [inline engine/redis/pool.go:77]:
```suggestion
		if err := proto.Unmarshal([]byte(val), res); err != nil {
			return nil, 0, err
```

File: storage/manager.go [hunk 7]
```diff
@@ -155,19 +153,19 @@ func (m *Manager) AddPool(name string, pool engine.Engine, threshold int64) {
 	}()
 }
 
-func (m *Manager) AddJob(ctx context.Context, job *model.JobData) error {
+func (m *Manager) AddJob(ctx context.Context, job engine.Job, poolName string) error {
```
Comment (git-hulk) [inline storage/manager.go:156]: Put the param `poolName` before `job`

File: storage/persistence/spanner/spanner.go [hunk 8]
```diff
@@ -45,15 +46,31 @@ func NewSpanner(cfg *config.SpannerConfig) (*Spanner, error) {
 }
 
 // BatchAddJobs write jobs data into secondary storage
-func (s *Spanner) BatchAddJobs(ctx context.Context, jobs []*model.JobData) (err error) {
+func (s *Spanner) BatchAddJobs(ctx context.Context, jobs []engine.Job, poolName string) (err error) {
```
Comment (git-hulk) [inline storage/persistence/spanner/spanner.go:49]: Put the param `poolName` before `jobs`

File: engine/redis_v2/pool.go [hunk 9]
```diff
@@ -68,7 +72,13 @@ func (p *Pool) Get(namespace, queue, jobID string) (body []byte, ttlSecond uint3
 			ttl = 0
 		}
 		metrics.poolGetJobs.WithLabelValues(p.redis.Name).Inc()
-		return []byte(val), uint32(ttl), nil
+
+		res := &model.JobData{}
+		err = proto.Unmarshal([]byte(val), res)
```
Comment (git-hulk) [inline engine/redis_v2/pool.go:77]: Unmarshal data should be outside the pool

**Comment by Napoleon0621** [comment]
Hi @git-hulk, just made a new commit to resolve the previous comments. 
Please have a look, thx.

**Review by git-hulk — APPROVED** [review]

**Pull request merged**